encoder/libav_encoder.cpp

```diff
@@ -0,0 +1,548 @@
+	if (!frame)
+		throw std::runtime_error("could not allocate AVFrame");
+
+	if (!video_start_ts_)
+		video_start_ts_ = timestamp_us;
+
+	frame->format = AV_PIX_FMT_DRM_PRIME;
+	frame->width = info.width;
+	frame->height = info.height;
+	frame->linesize[0] = info.stride;
+	frame->pts = timestamp_us - video_start_ts_ + (options_->av_sync < 0 ? -options_->av_sync : 0);
+
+	frame->buf[0] = av_buffer_alloc(sizeof(AVDRMFrameDescriptor));
+	frame->data[0] = frame->buf[0]->data;
+
+	AVDRMFrameDescriptor *desc = (AVDRMFrameDescriptor *)frame->data[0];
+	desc->nb_objects = 1;
+	desc->objects[0].fd = fd;
+	desc->objects[0].size = size;
+	desc->objects[0].format_modifier = DRM_FORMAT_MOD_INVALID;
+
+	desc->nb_layers = 1;
+	desc->layers[0].format = DRM_FORMAT_YUV420;
+	desc->layers[0].nb_planes = 3;
+	desc->layers[0].planes[0].object_index = 0;
+	desc->layers[0].planes[0].offset = 0;
+	desc->layers[0].planes[0].pitch = info.stride;
+	desc->layers[0].planes[1].object_index = 0;
+	desc->layers[0].planes[1].offset = info.stride * info.height;
+	desc->layers[0].planes[1].pitch = info.stride >> 1;
+	desc->layers[0].planes[2].object_index = 0;
+	desc->layers[0].planes[2].offset = info.stride * info.height * 5 / 4;
+	desc->layers[0].planes[2].pitch = info.stride >> 1;
+
+	std::scoped_lock<std::mutex> lock(video_mutex_);
+	frame_queue_.push(frame);
+	video_cv_.notify_all();
+}
+
+void LibAvEncoder::initOutput()
+{
+	int ret;
+
+	// Copy the global header from the video encode context once the first frame
+	// has been encoded.
+	avcodec_parameters_from_context(stream_[Video]->codecpar, codec_ctx_[Video]);
+
+	char err[64];
+	if (!(out_fmt_ctx_->flags & AVFMT_NOFILE))
+	{
+		ret = avio_open2(&out_fmt_ctx_->pb, options_->output.c_str(), AVIO_FLAG_WRITE, nullptr, nullptr);
+		if (ret < 0)
+		{
+			av_strerror(ret, err, sizeof(err));
+			throw std::runtime_error("libav: unable to open output mux for " + options_->output + ": " + err);
+		}
+	}
+
+	ret = avformat_write_header(out_fmt_ctx_, nullptr);
+	if (ret < 0)
+	{
+		av_strerror(ret, err, sizeof(err));
+		throw std::runtime_error("libav: unable write output mux header for " + options_->output + ": " + err);
+	}
+}
+
+void LibAvEncoder::deinitOutput()
+{
+	if (!out_fmt_ctx_)
+		return;
+
+	av_write_trailer(out_fmt_ctx_);
+
+	if (!(out_fmt_ctx_->flags & AVFMT_NOFILE))
+		avio_closep(&out_fmt_ctx_->pb);
+}
+
+void LibAvEncoder::encode(AVPacket *pkt, unsigned int stream_id)
+{
+	int ret = 0;
+
+	while (ret >= 0)
+	{
+		ret = avcodec_receive_packet(codec_ctx_[stream_id], pkt);
+
+		if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF)
+		{
+			av_packet_unref(pkt);
+			break;
+		}
+		else if (ret < 0)
+			throw std::runtime_error("libav error receiving packet: " + std::to_string(ret));
+
+		// Initialise the ouput mux on the first received video packet, as we may need
+		// to copy global header data from the encoder.
+		if (stream_id == Video && !init_output_)
+		{
+			initOutput();
+			init_output_ = true;
```

Comment:
`init_output_ = true` sounds a bit like the output still needs to be init-ed, but I realise I'm splitting hairs and probably being very annoying.

Reply:
I'll rename to output_ready_ :)